src/main.c

```diff
@@ -80,6 +100,129 @@ make_conn(const char *hostname)
   return -1;
 }
 
+/*
+ * Copy all the ciphertext bytes from buf into the client session.
+ * Returns 0 for success, 1 for error.
+ */
+int
+copy_tls_bytes_into_client_session(
+  struct rustls_client_session *client_session, uint8_t *buf, size_t len)
+{
+  ssize_t n;
+  int result;
+
+  while(len > 0) {
+    n = rustls_client_session_read_tls(client_session, (uint8_t *)buf, len);
+    if(n == 0) {
+      fprintf(stderr, "EOF from ClientSession::read_tls %ld \n", len);
+      return 1;
+    }
+    else if(n < 0) {
+      fprintf(stderr, "Error in ClientSession::read_tls\n");
+      return 1;
+    }
+    if((size_t)n > len) {
+      fprintf(stderr, "too many bytes written to ClientSession; overflow\n");
+      return 1;
+    }
+    len -= n;
+    buf += n;
+
+    result = rustls_client_session_process_new_packets(client_session);
+    if(result != RUSTLS_RESULT_OK) {
+      fprintf(stderr, "Error in process_new_packets\n");
+      return 1;
+    }
+  }
+  return 0;
+}
+
+/* Read all available bytes from the client_session until EOF.
+ * Note that EOF here indicates "no more bytes until
+ * process_new_packets", not "stream is closed".
+ *
+ * Returns 0 for success, 1 for error.
+ */
+int
+copy_plaintext_to_stdout(struct rustls_client_session *client_session)
+{
+  int result = 1;
+  char buf[2048];
+  int n;
+
+  for(;;) {
+    bzero(buf, sizeof(buf));
+    n =
+      rustls_client_session_read(client_session, (uint8_t *)buf, sizeof(buf));
+    if(n == 0) {
+      /* EOF from ClientSession::read. This is expected. */
+      return 0;
+    }
+    else if(n < 0) {
+      fprintf(stderr, "Error in ClientSession::read\n");
+      return 1;
+    }
+
+    result = write_all(STDOUT_FILENO, buf, n);
+    if(result != 0) {
+      return result;
+    }
+  }
+
+  return 0;
```

Comment:
nit: Is this `return` reachable? I wonder if some compilers might complain about unreachable code ... but then others might complain about a function without a `return`! Anyway, more constructively, since you never, ever expect to reach this point, you might reasonably return an error or even abort here to make it easier to debug such a strange condition.